packages/maestro-provider/src/api/api.ts

```diff
@@ -0,0 +1,164 @@
+import {
+	BadRequest,
+	type IElasticsearchService,
+	type IndexResult,
+	InternalServerError,
+	isEmpty,
+	logger,
+	type MaestroProviderConfig,
+	type RepositoryIndexingOperations,
+} from '@overture-stack/maestro-common';
+import { getRepoInformation, repository } from '@overture-stack/maestro-repository';
+
+/**
+ * Accumulates IndexResults objects, combining their failure data and determining overall success
+ * @param accumulator The initial `IndexResult` object that serves as the base for merging.
+ * @param result The `IndexResult` object to merge with the accumulator
+ * @returns
+ */
+const mergeResult = (accumulator: IndexResult, result: IndexResult): IndexResult => {
+	return {
+		indexName: result.indexName,
+		failureData: { ...accumulator.failureData, ...result.failureData },
+		successful: Object.keys(accumulator.failureData).length === 0 && Object.keys(result.failureData).length === 0,
+	};
+};
+
+/**
+ * Creates an object containing indexing operations to be used in the API
+ * @param config The configuration object for the `MaestroProvider`, which includes repository information
+ * @param indexer An implementation of `IElasticsearchService` used for performing Elasticsearch operations
+ * @returns
+ */
+export const api = (config: MaestroProviderConfig, indexer: IElasticsearchService): RepositoryIndexingOperations => {
+	const repositories = config.repositories;
+	if (!repositories) {
+		return {
+			indexOrganization: () => {
+				throw new InternalServerError(`Invalid repository configuration`);
+			},
+			indexRecord: () => {
+				throw new InternalServerError(`Invalid repository configuration`);
+			},
+			indexRepository: () => {
+				throw new InternalServerError(`Invalid repository configuration`);
+			},
+			removeIndexRecord: () => {
+				throw new InternalServerError(`Invalid repository configuration`);
+			},
+		};
+	}
+
+	/**
+	 * Performs asynchronous fetch and indexing operations for a specified repository
+	 * @param repoCode
+	 * @returns
+	 */
+	const indexRepository = async (repoCode: string): Promise<IndexResult> => {
+		const repoInfo = getRepoInformation(repositories, repoCode);
+
+		if (!repoInfo) {
+			logger.error(`Invalid repository information for repository code '${repoCode}'`);
+			throw new BadRequest(`Invalid repository code '${repoCode}'`);
+		}
```

Comment:
Would you consider returning `Result` objects for these instead of throwing errors?

On success they would return an object like:
```ts
{
  success: true,
  data: IndexResult
}
```

And on an error they would return:
```ts
{
  success: false,
  reason: 'InvalidArgument',
  message: `Invalid repository code '${repoCode}'`
}
```

The advantage here is that we can define the response types so that the caller of the API is informed of all possible outcomes and can use this information when handling the api action. The caller should still have a try/catch wrapper to protect their application, but it should only catch errors that aren't expected. The point is to communicate to the caller what outcomes they should expect, and errors are not communicated or enforced through any TS tooling.